editor/components/post-permalink/index.js

```diff
@@ -8,61 +8,88 @@ import { connect } from 'react-redux';
  */
 import { Component } from '@wordpress/element';
 import { __ } from '@wordpress/i18n';
-import { Dashicon, ClipboardButton, Button } from '@wordpress/components';
+import { Dashicon, Button, ClipboardButton, Tooltip } from '@wordpress/components';
 
 /**
  * Internal Dependencies
  */
 import './style.scss';
-import { isEditedPostNew, getEditedPostAttribute } from '../../store/selectors';
+import PostPermalinkEditor from './editor.js';
+import { isEditedPostNew, isPermalinkEditable, getEditedPostPreviewLink, getSamplePermalink } from '../../store/selectors';
+import { getWPAdminURL } from '../../utils/url';
 
 class PostPermalink extends Component {
 	constructor() {
 		super( ...arguments );
+
 		this.state = {
-			showCopyConfirmation: false,
+			iconClass: '',
+			editingPermalink: false,
 		};
-		this.onCopy = this.onCopy.bind( this );
-		this.onFinishCopy = this.onFinishCopy.bind( this );
-	}
-
-	componentWillUnmount() {
-		clearTimeout( this.dismissCopyConfirmation );
 	}
 
-	onCopy() {
-		this.setState( {
-			showCopyConfirmation: true,
-		} );
-	}
-
-	onFinishCopy() {
-		this.setState( {
-			showCopyConfirmation: false,
-		} );
+	componentDidUpdate( prevProps, prevState ) {
+		// If we've just stopped editing the permalink, focus on the new permalink.
+		if ( prevState.editingPermalink && ! this.state.editingPermalink ) {
+			const permalinkButton = document.querySelector( '.editor-post-permalink__link' );
```

Comment:
`document.querySelector` is almost never a good idea within a component, as we have no way to guarantee that via componentization, there's a single instance of the matching selector.

Typically, if a DOM node is needed, you could assign a [`ref`](https://reactjs.org/docs/refs-and-the-dom.html) to assign the node as an instance variable. This is made a bit more complex because `Button` is a component, and the `ref` would be the instance of the component, not the DOM node. This can be overcome by using [`React.findDOMNode( ref )`](https://reactjs.org/docs/react-dom.html#finddomnode), but also tends to be discouraged for reasons enumerated in the linked documentation.

Here, I'm also curious how we landed at this being the desired behavior. When I click "OK", it's a bit unexpected that focus is forced off the OK button to the URL button. Obviously we'd not want keyboard focus to be simply disappear when completing the edits, either by keyboard pressing the OK button or pressing escape (this isn't implemented, but should it be?).

To this end, a couple alternatives exist:

- [`withFocusReturn`](https://github.com/WordPress/gutenberg/blob/master/components/higher-order/with-focus-return/index.js) (apparently lacking documentation) is intended to deal with these scenarios where a component is intentionally short-lived, like with the permalink editor, returning focus when the component unmounts.
- We only lose focus because the "Edit" and "OK" buttons are not a shared element so far as React's DOM reconciler is concerned. It might take some restructuring, but I see no reason to think this _shouldn't_ be the same; it's effectively a single toggle between the two states of editing.
   - Example: https://codepen.io/aduth/pen/wmyGLB
   - One potential downside here is that the confirmation button should semantically be a `submit` within the `form`, but a `form` should only be rendered when editing.

Reply:
I went with that, as it seemed like the most logical point to focus on: the user has just edited the permalink, so focus on the permalink when they're done.

I went with adding a `focus()` method to `Button`, so it can handle its own focus behaviour.